### PR TITLE
Adding a callback buffer type

### PR DESCRIPTION
**Pull request by shangjiaxuan**
#325 I've made a few changes to make this functionality. Looked up the buffer APIs and added a few checks for incompatible calls. The Gen and Delete functions for the buffer objects currently uses placement new and delete, so as long as the constructor initializes the function pointers to null, the buffer will be in a valid state.

If there's problems with this, I will look into it later. Thanks!

### CODE REVIEW

**Comment by kcat** [comment]
It doesn't look like there's any changes here. Just a merge commit with nothing in it.

**Comment by shangjiaxuan** [comment]
I'm really sorry that I forgot to merge back the working branch. I will also clean up the naming conventions right away...

File: al/source.cpp [hunk 1]
```diff
@@ -3301,8 +3304,12 @@ ALsource::~ALsource()
...
         std::unique_ptr<ALbufferlistitem> head{BufferList};
-        BufferList = head->mNext.load(std::memory_order_relaxed);
-        if(ALbuffer *buffer{head->mBuffer}) DecrementRef(buffer->ref);
+		/* If (BufferList->next==nullptr), the following will segfault, 
+		* thus the loading of next is moved to the next line, hopefully
+		* this won't break anything.
+		*/
+        if(ALbuffer *buffer{BufferList->mBuffer}) DecrementRef(buffer->ref);
+		BufferList = head->mNext.load(std::memory_order_relaxed);
     }
```
Comment (shangjiaxuan) [inline al/source.cpp]: This is not needed with  the `if(ALbuffer *buffer{head->mBuffer}) DecrementRef(buffer->ref);`

File: al/source.cpp [hunk 2]
```diff
@@ -3301,8 +3304,12 @@ ALsource::~ALsource()
     while(BufferList != nullptr)
     {
         std::unique_ptr<ALbufferlistitem> head{BufferList};
-        BufferList = head->mNext.load(std::memory_order_relaxed);
-        if(ALbuffer *buffer{head->mBuffer}) DecrementRef(buffer->ref);
+		/* If (BufferList->next==nullptr), the following will segfault, 
+		* thus the loading of next is moved to the next line, hopefully
+		* this won't break anything.
+		*/
+        if(ALbuffer *buffer{BufferList->mBuffer}) DecrementRef(buffer->ref);
```
Comment (shangjiaxuan) [inline al/source.cpp]: This is not needed with  the `if(ALbuffer *buffer{head->mBuffer}) DecrementRef(buffer->ref);`

**Comment by kcat** [comment]
The main problem I see is that this won't work with non-mono formats since it's running the callback in the per-channel loop (so each channel of the format will invoke the callback separately). The callback will need to be invoked after the number of source samples is determined, before working on each channel.

There would also need to be some protection that the buffer can't be queued onto a source, and probably can't be set on multiple sources at once.

Also I'm not yet sure the best way to deal with the resampler padding. Always using all the stored samples would essentially create a constant delay, which would cause a problem for apps that want synchronization between callback and non-callback sources. I think what would be needed here is some kind of "first mix" flag that indicates it needs it get extra samples from the callback to fill in the resampler padding, and then with subsequent mixes it can restore all the padding samples and only get new samples from the callback (the `VOICE_IS_FADING` flag kind of already serves this purpose, in reverse).

It might also be a good idea to make it so when the callback provides less than the requested number of samples, it makes the source stop.

**Comment by shangjiaxuan** [comment]
A thought on stereo output is to use planar output for stereo or more channels (more arguments), and moving the buffer load of callback outside the loop seems like a valid solution, but that would require more intrusive modifications.

You are right about stopping the source when buffer ends is a good idea, I will look into that. For the re-sampler padding thing I think it's just a matter of offset in the load. I think it may be implemented by calling the callback as soon as it is created to restore the offset.

**Comment by kcat** [comment]
> A thought on stereo output is to use planar output for stereo or more channels (more arguments)

I'm not sure that's a good idea since currently all buffer data is expected to be interleaved channels. Planar layouts for multi-channel formats would be a good idea, but it would be specified by the caller ahead of time (likely as buffer properties, allowing the app to more generally handle padding between channels) and be independent of whether it's a callback or not.

> But still, you will need to know the following samples to actually do the re-sampling. Thus it will always be off by the padding to the future, and it seems a flag cannot really fix the inherent need to know some samples ahead.

In a sense, that's true. The callback will always have sent samples ahead of the current playback position. But the idea is that a source with a normal buffer and a source with a callback buffer can be started at the same time (with `alSourcePlayv`), and if the callback provides the exact same sample data the normal buffer has, the two should be heard perfectly synchronized.

> Also the audio thread works on a time resolution of each callback from the underlying back-end. Any time that is smaller is supposed to be dealt with implicitly (aka, works like a buffer stream with the exact buffer size needed, with no queuing and mapped memory).

I'm not sure what you mean. But as far as the backend is concerned, the update size isn't fixed. Some backends, like PulseAudio and WASAPI, will mix however many samples it can when it's told to. But even other backends, like ALSA or DSound, will mix a multiple of the period size if there's enough space to. It should not try to mix less than one update/period size, that much is guaranteed, but it can mix anything equal to or greater than one update/period. Plus, given the sample rate of the buffer and any pitch shifting on the source, the mixer may need to break up a mix into multiple partial mixes so it doesn't overflow certain internal buffers.

**Comment by kcat** [comment]
Apologies for this, but I got inspired yesterday and worked out an implementation that should work correctly with synchronization, multi-channel formats, and stopping at the end. I haven't gotten around to testing it yet, and the extension isn't finalized, so if you have any suggestions for changes or you want to get your example working with it, you're more than welcome.

**Comment by shangjiaxuan** [comment]
I will look at this now. My test is not really comprehensive, though.

**Pull request closed without merging**